### ISSUE TITLE
[Issue]: Network bind errors for udp and ssdp servers

### ISSUE THREAD

**Issue by Xabis** [jellyfin/jellyfin#10002]
### Please describe your bug

Using latest master, a few network bindings fail in Windows. These errors seem related to network discovery, where a multicast/broadcast binding is in play.

I am running the server directly inside VS 2020, in debug mode, and it should be noted that the current official windows 10.8.10 build has none of these issues on the same machine.

With a fresh/empty data directory, opening the server via the jellyfin-web client does not automatically start the initial setup. Instead a server list is presented with the hostname of the machine. Clicking "Connect" on this server fails, displaying an "Unable to connect" modal dialog.

The only way around the connectivity issue is by adding a new server, using a direct ip address (ex: 127.0.0.1), and then removing the automatically provided server. Once the page is reloaded, the first time setup is presented and may be completed as normal.

It is assumed that the connectivity via "server name" is due to these binding failures in the discovery protocol.

A point of interest is that turning off multi socket binding seems to prevent the binding error:
```
public UdpServerEntryPoint(...)
{
   _enableMultiSocketBinding = false; // OperatingSystem.IsWindows() || OperatingSystem.IsLinux();
}
```
```
var enableMultiSocketBinding = false; // OperatingSystem.IsWindows() || OperatingSystem.IsLinux();
_communicationsServer = new SsdpCommunicationsServer(_socketFactory, _networkManager, _logger, enableMultiSocketBinding)
{
    IsShared = true
};
```
However, additional failures are surfaced in the DLNA, and connecting via host still fails.
[12:02:07] [ERR] [6] Emby.Dlna.Main.DlnaEntryPoint: Failed to bind to multicast address: The requested address is not valid in its context.. DLNA will be unavailable

### Jellyfin Version

Other

### if other:

Latest master

### Environment

```markdown
- OS: Windows 11; VS 2020
- Clients: jellyfin-web (master)
- Browser: Firefox
```

### Jellyfin logs

```shell
[11:31:19] [WRN] [1] Emby.Server.Implementations.EntryPoints.UdpServerEntryPoint: Unable to start AutoDiscovery listener on UDP port 7359
System.Net.Sockets.SocketException (10049): The requested address is not valid in its context.
   at System.Net.Sockets.Socket.UpdateStatusAfterSocketErrorAndThrowException(SocketError error, Boolean disconnectOnFailure, String callerName)
   at System.Net.Sockets.Socket.DoBind(EndPoint endPointSnapshot, SocketAddress socketAddress)
   at System.Net.Sockets.Socket.Bind(EndPoint localEP)
   at Emby.Server.Implementations.Udp.UdpServer.Start(CancellationToken cancellationToken) in D:\dev\src\jellyfin\Emby.Server.Implementations\Udp\UdpServer.cs:line 91
   at Emby.Server.Implementations.EntryPoints.UdpServerEntryPoint.RunAsync() in D:\dev\src\jellyfin\Emby.Server.Implementations\EntryPoints\UdpServerEntryPoint.cs:line 87
[11:31:19] [INF] [1] Emby.Server.Implementations.ApplicationHost: Executed all post-startup entry points in 0:00:20.3895185
[11:31:19] [ERR] [15] Emby.Dlna.Main.DlnaEntryPoint: Error in CreateMulticastSocketsAndListen. IP address: 127.0.0.1
System.Net.Sockets.SocketException (10049): The requested address is not valid in its context.
   at System.Net.Sockets.Socket.UpdateStatusAfterSocketErrorAndThrowException(SocketError error, Boolean disconnectOnFailure, String callerName)
   at System.Net.Sockets.Socket.DoBind(EndPoint endPointSnapshot, SocketAddress socketAddress)
   at System.Net.Sockets.Socket.Bind(EndPoint localEP)
   at Emby.Server.Implementations.Net.SocketFactory.CreateUdpMulticastSocket(IPAddress multicastAddress, IPData bindInterface, Int32 multicastTimeToLive, Int32 localPort) in D:\dev\src\jellyfin\Emby.Server.Implementations\Net\SocketFactory.cs:line 94
   at Rssdp.Infrastructure.SsdpCommunicationsServer.CreateMulticastSocketsAndListen() in D:\dev\src\jellyfin\RSSDP\SsdpCommunicationsServer.cs:line 362
[11:31:19] [INF] [1] Main: Startup complete 0:00:23.2992367
[11:31:19] [INF] [16] Emby.Server.Implementations.ScheduledTasks.TaskManager: TaskCleanCollections Completed after 0 minute(s) and 0 seconds
[11:31:19] [ERR] [15] Emby.Dlna.Main.DlnaEntryPoint: Error in CreateMulticastSocketsAndListen. IP address: 192.168.1.2
System.Net.Sockets.SocketException (10049): The requested address is not valid in its context.
   at System.Net.Sockets.Socket.UpdateStatusAfterSocketErrorAndThrowException(SocketError error, Boolean disconnectOnFailure, String callerName)
   at System.Net.Sockets.Socket.DoBind(EndPoint endPointSnapshot, SocketAddress socketAddress)
   at System.Net.Sockets.Socket.Bind(EndPoint localEP)
   at Emby.Server.Implementations.Net.SocketFactory.CreateUdpMulticastSocket(IPAddress multicastAddress, IPData bindInterface, Int32 multicastTimeToLive, Int32 localPort) in D:\dev\src\jellyfin\Emby.Server.Implementations\Net\SocketFactory.cs:line 94
   at Rssdp.Infrastructure.SsdpCommunicationsServer.CreateMulticastSocketsAndListen() in D:\dev\src\jellyfin\RSSDP\SsdpCommunicationsServer.cs:line 362
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by Shadowghost** [comment]
Since I can't test this locally: Does everything work, if you set `_enableMultiSocketBinding = OperatingSystem.IsLinux();` and `var enableMultiSocketBinding = OperatingSystem.IsLinux();` or are there any additional errors (especially in regards to DLNA and IPv4/IPv6 addresses on the same interface)?

**Comment by Xabis** [comment]
@Shadowghost, no unfortunately even with these two properties set to false, the bind fails in RSSDP\SsdpCommunicationsServer.cs @ line 374:

`var socket = _SocketFactory.CreateUdpMulticastSocket(multicastGroupAddress, new IPData(IPAddress.Any, null), _MulticastTtl, SsdpConstants.MulticastPort);`

```
System.Net.Sockets.SocketException
  HResult=0x80004005
  Message=The requested address is not valid in its context.
  Source=System.Net.Sockets
  StackTrace:
   at System.Net.Sockets.Socket.UpdateStatusAfterSocketErrorAndThrowException(SocketError error, Boolean disconnectOnFailure, String callerName) in /_/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:line 3650
   at System.Net.Sockets.Socket.DoBind(EndPoint endPointSnapshot, SocketAddress socketAddress) in /_/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:line 788
   at System.Net.Sockets.Socket.Bind(EndPoint localEP) in /_/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:line 767
   at Emby.Server.Implementations.Net.SocketFactory.CreateUdpMulticastSocket(IPAddress multicastAddress, IPData bindInterface, Int32 multicastTimeToLive, Int32 localPort) in D:\dev\src\jellyfin\Emby.Server.Implementations\Net\SocketFactory.cs:line 94
   at Rssdp.Infrastructure.SsdpCommunicationsServer.CreateMulticastSocketsAndListen() in D:\dev\src\jellyfin\RSSDP\SsdpCommunicationsServer.cs:line 374
   at Rssdp.Infrastructure.SsdpCommunicationsServer.BeginListeningForMulticast() in D:\dev\src\jellyfin\RSSDP\SsdpCommunicationsServer.cs:line 120

  This exception was originally thrown at this call stack:
    System.Net.Sockets.Socket.UpdateStatusAfterSocketErrorAndThrowException(System.Net.Sockets.SocketError, bool, string) in Socket.cs
    System.Net.Sockets.Socket.DoBind(System.Net.EndPoint, System.Net.Internals.SocketAddress) in Socket.cs
    System.Net.Sockets.Socket.Bind(System.Net.EndPoint) in Socket.cs
    Emby.Server.Implementations.Net.SocketFactory.CreateUdpMulticastSocket(System.Net.IPAddress, MediaBrowser.Model.Net.IPData, int, int) in SocketFactory.cs
    Rssdp.Infrastructure.SsdpCommunicationsServer.CreateMulticastSocketsAndListen() in SsdpCommunicationsServer.cs
    Rssdp.Infrastructure.SsdpCommunicationsServer.BeginListeningForMulticast() in SsdpCommunicationsServer.cs
```

**Comment by cvium** [comment]
In my testing and research, it appears that Windows requires binding to a local IP address. You can still have the socket only listening to broadcasts from specific IPs, but the socket must be bound to local IP.

One such option is to use the unicast address on the relevant NICs.

**Comment by Sky-High** [comment]
FWIW: my system shows the same errors.

In line with the suggestion of @cvium the two culprit bind statements seem to be
https://github.com/jellyfin/jellyfin/blob/043fc387e08d5828864c8dcf4f5bcc2a64ecf761/Emby.Server.Implementations/Net/SocketFactory.cs#L94
and
https://github.com/jellyfin/jellyfin/blob/043fc387e08d5828864c8dcf4f5bcc2a64ecf761/Emby.Server.Implementations/Udp/UdpServer.cs#L91

Commit https://github.com/jellyfin/jellyfin/commit/42498194d9a4069b8cdeb9446f2714f74e3169de changed the bindings from "local address" 0.0.0.0:7359 (IPAddress.Any) to broadcast addresses
255.255.255.255:7359, 127.255.255.255:7359, 192.168.1.255:7359,   

And https://github.com/jellyfin/jellyfin/commit/b01d169d28cb7cfffa33796dfe7bf8be5570593a#diff-045dfc3f1c5db1ea605cd447cda0bbe15567f3541c25f7e32a5d5c37fea2750a changed the bindings from local addresses 127.0.0.1:1900, 192.168.1.xxx:1900 etc to multicast address 239.255.255.250:1900.

Maybe using those broadcast / multicast addresses is in theory better, but in Windows apparently this does not work. Maybe it does in Unix?

Anyhow, reverting those "culprit" statements to the pre-commit state:
https://github.com/jellyfin/jellyfin/blob/043fc387e08d5828864c8dcf4f5bcc2a64ecf761/Emby.Server.Implementations/Net/SocketFactory.cs#L94
to
`socket.Bind(new IPEndPoint(bindIPAddress, localPort))`

and
https://github.com/jellyfin/jellyfin/blob/043fc387e08d5828864c8dcf4f5bcc2a64ecf761/Emby.Server.Implementations/Udp/UdpServer.cs#L91
to
`_udpSocket.Bind(new IPEndPoint(IPAddress.Any, _endpoint.Port));`

removes the error messages, and TCPview shows that jellyfin is listening again on ports 1900 and 7359.

This does not mean that everything now works as intended, because for instance in the broadcast case port 7359, the socket will now receive data from any, potentially spurious, address reaching the network interface, and the code probably has to check somehow that the data was indeed sent to a valid broadcast address when receivingthe data. I am not sure if the pre-commit code indeed performed those checks.

Just my two cents...

**Comment by Shadowghost** [comment]
I'm aware of the problems and working on them, just don't have the time right now.
The main problem is Windows apparently isn't able to bind to addresses not existing on the host, even though they are proper broadcast or multicast addresses while it works fine on Linux.

**Comment by cvium** [comment]
Fixed by #10051
